### PR TITLE
LinalgRefactor-SGVector&SGMatrix-sum

### PR DESCRIPTION
**Pull request by OXPHOS**

### CODE REVIEW

File: src/shogun/mathematics/linalg/LinalgBackendEigen.h [hunk 1]
```diff
@@ -67,6 +67,28 @@ class LinalgBackendEigen : public LinalgBackendBase
...
+	BACKEND_GENERIC_SUM(int16_t);
+	BACKEND_GENERIC_SUM(uint16_t);
+	BACKEND_GENERIC_SUM(int32_t);
+	BACKEND_GENERIC_SUM(uint32_t);
+	BACKEND_GENERIC_SUM(int64_t);
+	BACKEND_GENERIC_SUM(uint64_t);
+	BACKEND_GENERIC_SUM(float32_t);
+	BACKEND_GENERIC_SUM(float64_t);
+	BACKEND_GENERIC_SUM(floatmax_t);
+	BACKEND_GENERIC_SUM(complex128_t);
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgBackendEigen.h]: btw do you actually try the operations with the complex type? or the int types?

File: src/shogun/mathematics/linalg/LinalgBackendViennacl.h [hunk 2]
```diff
@@ -115,6 +132,14 @@ class LinalgBackendViennaCL : public LinalgBackendGPUBase
 		return viennacl::linalg::inner_prod(a_gpu->data(a.size()), b_gpu->data(b.size()));
 	}
 
+	/** ViennaCL vector sum method */
+	template <typename T>
+	T sum_impl(const SGVector<T>& vec) const
+	{
+		GPUMemoryViennaCL<T>* vec_gpu=static_cast<GPUMemoryViennaCL<T>*>(vec.gpu_vector.get());
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgBackendViennacl.h]: as said, it would be good to make a private helper method for this and use it everywhere

**Comment by karlnapf** [comment]
Fine to merge from my side

File: src/shogun/mathematics/linalg/LinalgBackendEigen.h (hunk 1 above)
Comment (OXPHOS) [inline src/shogun/mathematics/linalg/LinalgBackendEigen.h]: i was using the int types. I didn't try complex types..

File: src/shogun/mathematics/linalg/LinalgBackendViennacl.h (hunk 2 above)
Comment (OXPHOS) [inline src/shogun/mathematics/linalg/LinalgBackendViennacl.h]: maybe i can send a separate patch after these patches? i'm afraid these minor changes will overload the travis.

Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgBackendViennacl.h]: absolutely, but do it from the next PR on

**Comment by OXPHOS** [comment]
@karlnapf @lambday SGVector and SGMatrix sum operation

File: src/shogun/mathematics/linalg/LinalgBackendEigen.h [hunk 3]
```diff
@@ -103,6 +114,20 @@ class LinalgBackendEigen : public LinalgBackendBase
...
+	/** Eigen3 vector sum method */
+	template <typename T>
+	T sum_impl(const SGVector<T>& vec) const
+	{
+		return (typename SGVector<T>::EigenVectorXtMap(vec)).sum();
+	}
+
+	/** Eigen3 matrix sum method */
+	template <typename T>
+	T sum_impl(const SGMatrix<T>& mat) const
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgBackendEigen.h]: can we please have row-wise and column wise sum here as well?
Should that be in a new patch or not?

**Comment by karlnapf** [comment]
Ill merge this one
We should definitely have one that slices in some direction

**Pull request merged**